File: src/main/java/cz/vhromada/catalog/gui/song/SongsPanel.java
package cz.vhromada.catalog.gui.song;

import javax.swing.JPanel;
import javax.swing.JTabbedPane;

import cz.vhromada.catalog.entity.Music;
import cz.vhromada.catalog.entity.Song;
import cz.vhromada.catalog.facade.SongFacade;
import cz.vhromada.catalog.gui.commons.AbstractInfoDialog;
import cz.vhromada.catalog.gui.commons.AbstractOverviewDataPanel;

import org.springframework.util.Assert;

/**
 * A class represents panel with songs' data.
 *
 * @author Vladimir Hromada
 */
public class SongsPanel extends AbstractOverviewDataPanel<Song> {

    /**
     * SerialVersionUID
     */
    private static final long serialVersionUID = 1L;

    /**
     * Facade for songs
     */
    private final SongFacade songFacade;

    /**
     * Music
     */
    private Music music;

    /**
     * Creates a new instance of SongsPanel.
     *
     * @param songFacade facade for songs
     * @param music      music
     * @throws IllegalArgumentException if facade for songs is null
     *                                  or music is null
     */
    public SongsPanel(final SongFacade songFacade, final Music music) {
        super(getSongsListDataModel(songFacade, music));

        this.songFacade = songFacade;
        this.music = music;
    }

    /**
     * Sets a new value to music.
     *
     * @param music new value
     * @throws IllegalArgumentException if music is null
     */
    public void setMusic(final Music music) {
        Assert.notNull(music, "Music mustn't be null.");

        this.music = music;
    }

    @Override
    public void newData() {
        throw new UnsupportedOperationException("Creating new data is not allowed for songs.");
    }

    @Override
    public void clearSelection() {
        throw new UnsupportedOperationException("Clearing selection is not allowed for songs.");
    }

    @Override
    public void save() {
        throw new UnsupportedOperationException("Saving data is not allowed for songs.");
    }

    @Override
    public boolean isSaved() {
        throw new UnsupportedOperationException("Testing if data are saved is not allowed for songs.");
    }

    @Override
    protected AbstractInfoDialog<Song> getInfoDialog(final boolean add, final Song data) {
        return add ? new SongInfoDialog() : new SongInfoDialog(data);
    }

    @Override
    protected void addData(final Song data) {
        songFacade.add(music, data);
    }

    @Override
    protected void deleteData() {
        throw new UnsupportedOperationException("Deleting data is not allowed for songs.");
    }

    @Override
    protected void updateData(final Song data) {
        songFacade.update(data);
    }

    @Override
    protected void removeData(final Song data) {
        songFacade.remove(data);
    }

    @Override
    protected void duplicatesData(final Song data) {
        songFacade.duplicate(data);
    }

    @Override
    protected void moveUpData(final Song data) {
        songFacade.moveUp(data);
    }

    @Override
    protected void moveDownData(final Song data) {
        songFacade.moveDown(data);
    }

    @Override
    protected JPanel getDataPanel(final Song data) {
        return new SongDataPanel(data);
    }

    @Override
    protected void updateDataOnChange(final JTabbedPane dataPanel, final Song data) {
    }

    /**
     * Returns data model for list with songs.
     *
     * @param facade      facade for songs
     * @param musicObject music
     * @return data model for list with songs
     * @throws IllegalArgumentException if facade for songs is null
     *                                  or music is null
     */
    private static SongsListDataModel getSongsListDataModel(final SongFacade facade, final Music musicObject) {
        return new SongsListDataModel(facade, musicObject);
    }

}

File: src/main/kotlin/cz/vhromada/catalog/gui/song/SongsPanel.kt
package cz.vhromada.catalog.gui.song

import cz.vhromada.catalog.entity.Music
import cz.vhromada.catalog.entity.Song
import cz.vhromada.catalog.facade.SongFacade
import cz.vhromada.catalog.gui.common.AbstractInfoDialog
import cz.vhromada.catalog.gui.common.AbstractOverviewDataPanel
import javax.swing.JPanel
import javax.swing.JTabbedPane

/**
 * A class represents panel with songs' data.
 *
 * @author Vladimir Hromada
 */
class SongsPanel(
        private val songFacade: SongFacade,
        private var music: Music) : AbstractOverviewDataPanel<Song>(SongsListDataModel(songFacade, music)) {

    /**
     * Sets a new value to music.
     *
     * @param music new value
     */
    fun setMusic(music: Music) {
        this.music = music
    }

    override fun newData() {
        error { "Creating new data is not allowed for songs." }
    }

    override fun clearSelection() {
        error { "Clearing selection is not allowed for songs." }
    }

    override fun save() {
        error { "Saving data is not allowed for songs." }
    }

    override fun getInfoDialog(add: Boolean, data: Song?): AbstractInfoDialog<Song> {
        return if (add) SongInfoDialog() else SongInfoDialog(data!!)
    }

    override fun addData(data: Song) {
        songFacade.add(music, data)
    }

    override fun deleteData() {
        error { "Deleting data is not allowed for songs." }
    }

    override fun updateData(data: Song) {
        songFacade.update(data)
    }

    override fun removeData(data: Song) {
        songFacade.remove(data)
    }

    override fun duplicatesData(data: Song) {
        songFacade.duplicate(data)
    }

    override fun moveUpData(data: Song) {
        songFacade.moveUp(data)
    }

    override fun moveDownData(data: Song) {
        songFacade.moveDown(data)
    }

    override fun getDataPanel(data: Song): JPanel {
        return SongDataPanel(data)
    }

    override fun updateDataOnChange(dataPanel: JTabbedPane, data: Song) {
        // nothing
    }

}
